release/releaser.go

```diff
@@ -45,539 +36,318 @@ func NewReleaser(
+	switch spec.ServiceSpec {
+	case flux.ServiceSpec(""):
+		ids := []flux.ServiceID{}
+		for _, s := range spec.ServiceSpecs {
+			if s == flux.ServiceSpecAll {
+				return rc.SelectServices(nil, lockedSet, excludedSet, results, logStatus)
 			}
-			return
-		}
-
-		status := flux.ReleaseStatusSuccess
-		if err != nil {
-			status = flux.ReleaseStatusFailed
-		}
-		// Filling this from the job is a temporary migration hack. Ideally all
-		// the release info should be stored on the release object in a releases
-		// table, and the job should really just have a pointer to that.
-		err2 = notifications.Release(cfg, flux.Release{
-			ID:        flux.ReleaseID(job.ID),
-			CreatedAt: job.Submitted,
-			StartedAt: job.Claimed,
-			// TODO: fetch the job and look this up so it matches
-			EndedAt:  time.Now().UTC(),
-			Done:     true,
-			Priority: job.Priority,
-			Status:   status,
-			Log:      job.Log,
-
-			Spec: flux.ReleaseSpec(job.Params.(jobs.ReleaseJobParams)),
-			// TODO: Populate this... It's just included so the user can't trigger a null-pointer.
-			Result: flux.ReleaseResult{},
-		}, err)
-		if err2 != nil {
-			if err == nil {
-				err = errors.Wrap(err2, "sending notifications")
+			id, err := flux.ParseServiceID(string(s))
+			if err != nil {
+				return nil, err
 			}
+			ids = append(ids, id)
 		}
-	}()
-
-	for i, action := range actions {
-		updateJob(action.Description)
-		inst.Log("description", action.Description)
-		if action.Do == nil {
-			continue
+		return rc.SelectExactServices(ids, lockedSet, excludedSet, results, logStatus)
+	case flux.ServiceSpecAll:
+		return rc.SelectServices(nil, lockedSet, excludedSet, results, logStatus)
+	default:
+		id, err := flux.ParseServiceID(string(spec.ServiceSpec))
+		if err != nil {
+			return nil, err
 		}
+		return rc.SelectExactServices([]flux.ServiceID{id}, lockedSet, excludedSet, results, logStatus)
+	}
+}
 
-		if kind == flux.ReleaseKindExecute {
-			begin := time.Now()
-			result, err := action.Do(rc)
-			r.metrics.ActionDuration.With(
-				fluxmetrics.LabelAction, action.Name,
-				fluxmetrics.LabelSuccess, fmt.Sprint(err == nil),
-			).Observe(time.Since(begin).Seconds())
-			if err != nil {
-				updateJob(err.Error())
-				inst.Log("err", err)
-				actions[i].Result = "Failed: " + err.Error()
-				return err
-			}
-			if result != "" {
-				updateJob(result)
-			}
-			actions[i].Result = result
+// Find all the image updates that should be performed, and do
+// replacements. (For a dry-run, we don't strictly need to do the
+// replacements, since we won't be committing any changes back;
+// however we do want to see if we *can* do the replacements, because
+// if not, it indicates there's likely some problem with the running
+// system vs the definitions given in the repo.)
+func calculateImageUpdates(inst *instance.Instance, updateable []*ServiceUpdate, spec *flux.ReleaseSpec, results flux.ReleaseResult, logStatus statusFn) ([]*ServiceUpdate, error) {
```

Comment:
Can I do this refactor in another PR please? I feel like I've rewritten enough (the status quo at least has the benefit of mostly leaving other packages alone)